File: Torplem.js
// Torplem.js
// Uma Biblioteca Matemática em JavaScripts
// Feito Por Caio Victor 2024

class Torplem {
    // Operações Básicas
    static soma(a, b) {
        return a + b;
    }

    static subtracao(a, b) {
        return a - b;
    }

    static multiplicacao(a, b) {
        return a * b;
    }

    static divisao(a, b) {
        if (b === 0) {
            throw new Error("Divisão por zero não é permitida");
        }
        return a / b;
    }

    // Propriedades dos Números
    static verificarParImpar(a) {
        return a % 2 === 0 ? "Par" : "Ímpar";
    }

    static verificarPositivoNegativoZero(a) {
        if (a > 0) return "Positivo";
        if (a < 0) return "Negativo";
        return "Zero";
    }

    // Operações Avançadas
    static fatorial(n) {
        if (n < 0) {
            throw new Error("Fatorial não está definido para números negativos");
        }
        if (n === 0) return 1;
        return Array.from({length: n}, (_, i) => i + 1).reduce((acc, val) => acc * val, 1);
    }

    static verificarPrimo(n) {
        if (n <= 1) return false;
        for (let i = 2; i <= Math.sqrt(n); i++) {
            if (n % i === 0) return false;
        }
        return true;
    }

    static potencia(base, expoente) {
        return Math.pow(base, expoente);
    }

    static raizQuadrada(n) {
        if (n < 0) {
            throw new Error("Raiz quadrada de número negativo não é real");
        }
        return Math.sqrt(n);
    }

    // Operações com Arrays
    static maximo(arr) {
        if (!arr.length) {
            throw new Error("Array vazio");
        }
        return Math.max(...arr);
    }

    static minimo(arr) {
        if (!arr.length) {
            throw new Error("Array vazio");
        }
        return Math.min(...arr);
    }

    static media(arr) {
        if (!arr.length) {
            throw new Error("Array vazio");
        }
        return arr.reduce((acc, val) => acc + val, 0) / arr.length;
    }

    // MDC e MMC
    static mdc(a, b) {
        while (b !== 0) {
            [a, b] = [b, a % b];
        }
        return Math.abs(a);
    }

    static mmc(a, b) {
        return Math.abs(a * b) / this.mdc(a, b);
    }

    // Conversão de Ângulos
    static grausParaRadianos(angulo) {
        return angulo * Math.PI / 180;
    }

    static radianosParaGraus(radianos) {
        return radianos * 180 / Math.PI;
    }

    // Números Especiais
    static verificarNumeroPerfeito(n) {
        if (n <= 0) return false;
        return Array.from({length: n-1}, (_, i) => i + 1)
            .filter(i => n % i === 0)
            .reduce((acc, val) => acc + val, 0) === n;
    }

    // Conversão de Base
    static converterBase(num, base) {
        if (base < 2 || base > 36) {
            throw new Error("Base deve estar entre 2 e 36");
        }
        return num.toString(base).toUpperCase();
    }

    // Funções Estatísticas
    static calcularMediana(arr) {
        if (!arr.length) {
            throw new Error("Array vazio");
        }
        const sorted = [...arr].sort((a, b) => a - b);
        const mid = Math.floor(sorted.length / 2);
        return sorted.length % 2 === 0
            ? (sorted[mid - 1] + sorted[mid]) / 2
            : sorted[mid];
    }

    static fibonacci(n) {
        if (n <= 0) return [];
        if (n === 1) return [1];
        const fib = [1, 1];
        for (let i = 2; i < n; i++) {
            fib[i] = fib[i-1] + fib[i-2];
        }
        return fib;
    }

    // Geração de Números Aleatórios
    static gerarNumeroAleatorio(min, max) {
        return Math.floor(Math.random() * (max - min + 1)) + min;
    }

    // Novas Funções
    static desvioPadrao(arr) {
        if (arr.length <= 1) {
            throw new Error("Array precisa ter pelo menos 2 elementos");
        }
        const media = this.media(arr);
        const soma = arr.reduce((acc, val) => acc + Math.pow(val - media, 2), 0);
        return Math.sqrt(soma / (arr.length - 1));
    }

    static verificarNumeroArmstrong(n) {
        if (n < 0) return false;
        const str = n.toString();
        const numDigitos = str.length;
        const soma = str
            .split('')
            .reduce((acc, digit) => acc + Math.pow(parseInt(digit), numDigitos), 0);
        return soma === n;
    }

    static areaCirculo(raio) {
        if (raio < 0) {
            throw new Error("Raio não pode ser negativo");
        }
        return Math.PI * Math.pow(raio, 2);
    }

    static volumeEsfera(raio) {
        if (raio < 0) {
            throw new Error("Raio não pode ser negativo");
        }
        return (4/3) * Math.PI * Math.pow(raio, 3);
    }

    static verificarNumeroTriangular(n) {
        if (n < 0) return false;
        let i = 1;
        let triangular = 0;
        while (triangular < n) {
            triangular += i;
            if (triangular === n) return true;
            i++;
        }
        return false;
    }

    static sequenciaCollatz(n) {
        if (n <= 0) {
            throw new Error("Número inicial deve ser positivo");
        }
        const sequencia = [n];
        while (n !== 1) {
            n = n % 2 === 0 ? n / 2 : 3 * n + 1;
            sequencia.push(n);
        }
        return sequencia;
    }
}

export default Torplem;

File: Torplem.py
# Torplem.py
# Uma Biblioteca Matemática em Python
# Feito Por Caio Victor 2024

import math
import random
from typing import List, Union, Tuple

class Torplem:
    @staticmethod
    def soma(a: float, b: float) -> float:
        return a + b

    @staticmethod
    def subtracao(a: float, b: float) -> float:
        return a - b

    @staticmethod
    def multiplicacao(a: float, b: float) -> float:
        return a * b

    @staticmethod
    def divisao(a: float, b: float) -> float:
        if b == 0:
            raise ValueError("Divisão por zero não é permitida")
        return a / b

    @staticmethod
    def verificar_par_impar(a: int) -> str:
        return "Par" if a % 2 == 0 else "Ímpar"

    @staticmethod
    def verificar_positivo_negativo_zero(a: float) -> str:
        if a > 0:
            return "Positivo"
        elif a < 0:
            return "Negativo"
        return "Zero"

    @staticmethod
    def fatorial(n: int) -> int:
        if n < 0:
            raise ValueError("Fatorial não está definido para números negativos")
        if n == 0:
            return 1
        return math.factorial(n)

    @staticmethod
    def verificar_primo(n: int) -> bool:
        if n <= 1:
            return False
        return all(n % i != 0 for i in range(2, int(math.sqrt(n)) + 1))

    @staticmethod
    def potencia(base: float, expoente: float) -> float:
        return math.pow(base, expoente)

    @staticmethod
    def raiz_quadrada(n: float) -> float:
        if n < 0:
            raise ValueError("Raiz quadrada de número negativo não é real")
        return math.sqrt(n)

    @staticmethod
    def maximo(arr: List[float]) -> float:
        if not arr:
            raise ValueError("Array vazio")
        return max(arr)

    @staticmethod
    def minimo(arr: List[float]) -> float:
        if not arr:
            raise ValueError("Array vazio")
        return min(arr)

    @staticmethod
    def media(arr: List[float]) -> float:
        if not arr:
            raise ValueError("Array vazio")
        return sum(arr) / len(arr)

    @staticmethod
    def mdc(a: int, b: int) -> int:
        return math.gcd(abs(a), abs(b))

    @staticmethod
    def mmc(a: int, b: int) -> int:
        return abs(a * b) // math.gcd(a, b)

    @staticmethod
    def graus_para_radianos(angulo: float) -> float:
        return math.radians(angulo)

    @staticmethod
    def radianos_para_graus(radianos: float) -> float:
        return math.degrees(radianos)

    @staticmethod
    def verificar_numero_perfeito(n: int) -> bool:
        if n <= 0:
            return False
        return sum(i for i in range(1, n) if n % i == 0) == n

    @staticmethod
    def converter_base(num: int, base: int) -> str:
        if not 2 <= base <= 36:
            raise ValueError("Base deve estar entre 2 e 36")
        if num == 0:
            return "0"
        
        digitos = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        negativo = num < 0
        num = abs(num)
        resultado = ""
        
        while num:
            resultado = digitos[num % base] + resultado
            num //= base
            
        return f"-{resultado}" if negativo else resultado

    @staticmethod
    def calcular_mediana(arr: List[float]) -> float:
        if not arr:
            raise ValueError("Array vazio")
        
        sorted_arr = sorted(arr)
        n = len(sorted_arr)
        mid = n // 2
        
        return (sorted_arr[mid] + sorted_arr[mid-1]) / 2 if n % 2 == 0 else sorted_arr[mid]

    @staticmethod
    def fibonacci(n: int) -> List[int]:
        if n <= 0:
            return []
        elif n == 1:
            return [1]
        
        fib = [1, 1]
        for i in range(2, n):
            fib.append(fib[i-1] + fib[i-2])
        return fib

    @staticmethod
    def gerar_numero_aleatorio(min_val: int, max_val: int) -> int:
        return random.randint(min_val, max_val)

    @staticmethod
    def desvio_padrao(arr: List[float]) -> float:
        if len(arr) <= 1:
            raise ValueError("Array precisa ter pelo menos 2 elementos")
        return math.sqrt(sum((x - Torplem.media(arr)) ** 2 for x in arr) / (len(arr) - 1))

    @staticmethod
    def verificar_numero_armstrong(n: int) -> bool:
        if n < 0:
            return False
        str_n = str(n)
        num_digitos = len(str_n)
        return sum(int(d) ** num_digitos for d in str_n) == n

    @staticmethod
    def area_circulo(raio: float) -> float:
        if raio < 0:
            raise ValueError("Raio não pode ser negativo")
        return math.pi * raio ** 2

    @staticmethod
    def volume_esfera(raio: float) -> float:
        if raio < 0:
            raise ValueError("Raio não pode ser negativo")
        return (4/3) * math.pi * raio ** 3

    @staticmethod
    def verificar_numero_triangular(n: int) -> bool:
        if n < 0:
            return False
        i = 1
        triangular = 0
        while triangular < n:
            triangular += i
            if triangular == n:
                return True
            i += 1
        return False

    @staticmethod
    def sequencia_collatz(n: int) -> List[int]:
        if n <= 0:
            raise ValueError("Número inicial deve ser positivo")
        sequencia = [n]
        while n != 1:
            n = n // 2 if n % 2 == 0 else 3 * n + 1
            sequencia.append(n)
        return sequencia